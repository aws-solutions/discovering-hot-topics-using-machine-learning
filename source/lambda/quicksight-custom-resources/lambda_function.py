#!/usr/bin/env python
######################################################################################################################
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0
######################################################################################################################

import logging

from crhelper import CfnResource
from shared_util import custom_logging

from util.quicksight import QuicksightApi

logger = custom_logging.get_logger(__name__)
helper = CfnResource(json_logging=False, log_level="INFO")


def get_resource_propertices(event):
    logger.debug(f"servicing request event:{event}")
    request_type = event["RequestType"]
    resource_properties = event["ResourceProperties"]
    resource = resource_properties["Resource"]

    logger.info(f"servicing request request_type:{request_type} resource:{resource}")
    return resource_properties


def log_exception(error):
    """
    Do logging in addition to crhelper exception handeling
    """
    logger.error(f"quicksight-application resources action error: {error}")
    import sys
    import traceback

    exc_type, exc_value, exc_traceback = sys.exc_info()
    format_exception = traceback.format_exception(exc_type, exc_value, exc_traceback)
    logger.error(repr(format_exception))


@helper.create
def custom_resource_create(event, _):
    process_create_request(event)
    return None


@helper.delete
def custom_resource_delete(event, _):
    process_delete_request(event)
    return None


@helper.update
def custom_resource_update(event, _):
    # For update we delete all the resources and re-create new ones. Any user customization on QuickSight may be lost
    try:
        process_delete_request(event)
        process_create_request(event)  # once deleted re-create all the resources
    except Exception as error:
        resource_properties = get_resource_propertices(event)
        qs_api = QuicksightApi(resource_properties)
        qs_api.delete_all_resources()
        # Do logging in addition to crhelper exception handeling
        log_exception(error)
        raise (error)

    return None


def process_create_request(event):
    request_type = event["RequestType"]  # "Create"
    resource_properties = get_resource_propertices(event)
    resource = resource_properties["Resource"]
    qs_api = QuicksightApi(resource_properties)

    try:
        if resource == "all":
            qs_api.create_all_resources()
        elif resource == "datasource":
            qs_api.create_data_source()
        elif resource == "dataset":
            qs_api.create_data_sets()
        elif resource == "analysis":
            qs_api.create_analysis()
        elif resource == "dashboard":
            qs_api.create_dashboard()
        else:
            logger.error(f"Not handling request resource:{resource}, request_type:{request_type}")
            raise ValueError(f"Received unsupported request request_type:{request_type}, resource:{resource}")
        if resource in ["all", "analysis", "dashboard"]:
            analysis_url = qs_api.quicksight_application.get_analysis().url
            dashboard_url = qs_api.quicksight_application.get_dashboard().url
            helper.Data.update({"analysis_url": analysis_url, "dashboard_url": dashboard_url})
    except Exception as error:
        # Do logging in addition to crhelper exception handeling
        log_exception(error)
        qs_api.delete_all_resources()
        raise (error)

    logger.info(f"finished with request_type:{request_type} resource:{resource}")
    return None


def process_delete_request(event):
    request_type = event["RequestType"]
    resource_properties = get_resource_propertices(event)
    resource = resource_properties["Resource"]
    qs_api = QuicksightApi(resource_properties)

    try:
        if resource == "all":
            qs_api.delete_all_resources()
        elif resource == "datasource":
            qs_api.delete_data_source()
        elif resource == "dataset":
            qs_api.delete_data_sets()
        elif resource == "analysis":
            qs_api.delete_analysis()
        elif resource == "dashboard":
            qs_api.delete_dashboard()
        else:
            logger.error(f"Not handling request resource:{resource}, request_type:{request_type}")
            raise ValueError(f"Received unsupported request request_type:{request_type}, resource:{resource}")
    except Exception as error:
        # Do logging in addition to crhelper exception handeling
        log_exception(error)
        raise (error)

    logger.info(f"finished with request_type:{request_type} resource:{resource}")
    return None


def handler(event, context):
    helper(event, context)
