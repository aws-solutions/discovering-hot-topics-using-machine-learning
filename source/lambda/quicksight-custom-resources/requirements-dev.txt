pytest-cov~=4.1.0
pytest-env~=1.0.1
moto~=4.2.4
pytest~=7.4.2
botocore
mock~=5.1.0
-e ../layers/python_lambda_layer