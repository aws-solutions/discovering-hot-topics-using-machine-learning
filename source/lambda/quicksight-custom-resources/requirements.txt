crhelper==2.0.11
requests==2.28.1
pyyaml
tenacity==8.1.0