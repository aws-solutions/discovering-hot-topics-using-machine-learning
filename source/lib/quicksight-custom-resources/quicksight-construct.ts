#!/usr/bin/env node
/**********************************************************************************************************************
 *  Copyright 2020-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                      *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { Effect, IRole, Policy, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

export enum QuickSightSetup {
    DATA_SET = 'dataset',
    DATA_SOURCE = 'datasource',
    ANALYSIS = 'analysis',
    DASHBOARD = 'dashboard',
    ALL = 'all'
}

export interface QuickSightProps {
    readonly name: string;
    readonly resource: QuickSightSetup;
    readonly sourceTemplateArn: string;
    readonly principalArn: string;
    readonly workgroupName: string;
    readonly logLevel: string;
    readonly role: IRole;
    readonly parentStackName: string;
}
export class QuickSight extends Construct {
    private _analysisURL: string;
    private _dashboardURL: string;

    constructor(scope: Construct, id: string, props: QuickSightProps) {
        super(scope, id);
        const qsCreateResource = this.createCustomResource(props);
        this._analysisURL = qsCreateResource.getAtt('analysis_url').toString();
        this._dashboardURL = qsCreateResource.getAtt('dashboard_url').toString();
    }

    private createCustomResource(props: QuickSightProps): cdk.CustomResource {
        const customResourcePolicy = new Policy(this, 'QSCustomResourcePolicy', {
            statements: [
                new PolicyStatement({
                    effect: Effect.ALLOW,
                    actions: [
                        'quicksight:CreateAnalysis',
                        'quicksight:DeleteAnalysis',
                        'quicksight:CreateDataSet',
                        'quicksight:DeleteDataSet',
                        'quicksight:CreateDataSource',
                        'quicksight:DeleteDataSource',
                        'quicksight:Describe*',
                        'quicksight:Get*',
                        'quicksight:List*',
                        'quicksight:PassDataSet',
                        'quicksight:PassDataSource',
                        'quicksight:RestoreAnalysis',
                        'quicksight:SearchAnalyses',
                        'quicksight:CreateDashboard',
                        'quicksight:DeleteDashboard'
                    ],
                    resources: [`arn:${cdk.Aws.PARTITION}:quicksight:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:*/*`]
                }),
                new PolicyStatement({
                    effect: Effect.ALLOW,
                    actions: ['quicksight:DescribeTemplate'],
                    resources: [props.sourceTemplateArn]
                })
            ]
        });

        customResourcePolicy.attachToRole(props.role);

        const customResourceFunction = new lambda.Function(this, 'CustomResource', {
            runtime: lambda.Runtime.PYTHON_3_8,
            handler: 'lambda_function.handler',
            role: props.role,
            code: lambda.Code.fromAsset('lambda/quicksight-custom-resources'),
            timeout: cdk.Duration.seconds(30)
        });
        customResourceFunction.node.addDependency(customResourcePolicy);

        (customResourceFunction.node.defaultChild as lambda.CfnFunction).addMetadata('cfn_nag', {
            rules_to_suppress: [
                {
                    'id': 'W89',
                    'reason': 'This is not a rule for the general case, just for specific use cases/industries'
                },
                {
                    'id': 'W92',
                    'reason': 'Impossible for us to define the correct concurrency for clients'
                }
            ]
        });

        const customResource = new cdk.CustomResource(this, 'QuickSightResources', {
            serviceToken: customResourceFunction.functionArn,
            properties: {
                Resource: props.resource,
                ApplicationName: props.name,
                StackName: props.parentStackName,
                LogLevel: props.logLevel,
                QuickSightSourceTemplateArn: props.sourceTemplateArn,
                QuickSightPrincipalArn: props.principalArn,
                WorkGroupName: props.workgroupName
            },
            resourceType: 'Custom::QuickSightResources'
        });

        customResource.node.addDependency(customResourcePolicy);
        return customResource;
    }

    public get analysisURL(): string {
        return this._analysisURL;
    }

    public get dashboardURL(): string {
        return this._dashboardURL;
    }
}
