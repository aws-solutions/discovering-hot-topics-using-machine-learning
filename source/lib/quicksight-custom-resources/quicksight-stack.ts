#!/usr/bin/env node
/**********************************************************************************************************************
* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
* SPDX-License-Identifier: Apache-2.0
 *********************************************************************************************************************/

import { CfnParameter, NestedStack, NestedStackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { ExecutionRole } from '../solution-helper/lambda-role-cloudwatch-construct';
import { QuickSight, QuickSightSetup } from './quicksight-construct';

export class QuickSightStack extends NestedStack {
    private _quickSight: QuickSight;

    constructor(scope: Construct, id: string, props:NestedStackProps) {
        super(scope, id, props);

        /**
         * The QuickSight source template arn should follow the format below. The sourceTemplateAccountId is the aws
         * account id that is sharing the template and should have
         * `arn:${cdk.Aws.PARTITION}:quicksight:${cdk.Aws.REGION}:${sourceTemplateAccountId}:template/${sourceTemplateName}`
         * either granted public access to its template or specific access for this account to use it.
         */
        const quickSightSourceTemplateArn = new CfnParameter(this, 'QuickSightSourceTemplateArn', {
            type: 'String',
            description: 'The Amazon QuickSight template arn to use as the source template when creating analysis and dashboard resources',
            allowedPattern: '^arn:\\S+:quicksight:\\S+:\\d{12}:template/\\S+$',
            constraintDescription: 'Provide an arn matching an Amazon Quicksight template arn. The input did not match the validation pattern.'
        });

        /**
         * The QuickSight principal is used as the principal in the permissions of QuickSight resources created by the custom resources.
         * `arn:${cdk.Aws.PARTITION}:quicksight:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:user/${namespace}/${user_name`
         *
         * Example for namescape is 'default'. You can use aws cli command quicksight list-users and list-namespaces to choose from
         * Sample CLI command:
         * aws quicksight list-namespaces --aws-account-id $AWS_ACCOUNT_ID
         * aws quicksight list-users --aws-account-id $AWS_ACCOUNT_ID  --namespace default
         *
         */
        const quicksightPrincipalArn = new CfnParameter(this, 'QuickSightPrincipalArn', {
            type: 'String',
            description: 'The Amazon QuickSight principal arn used in the permissions of QuickSight resources',
            allowedPattern: '^arn:\\S+:quicksight:\\S+:\\d{12}:user/\\S+$',
            constraintDescription: 'Provide an arn matching an Amazon Quicksight principal arn. The input did not match the validation pattern.'
        });

        /**
         * The Solution ID with which the nested stack is associated with
         */
        const solutionID = new CfnParameter(this, 'SolutionID', {
            type: 'String',
            description: 'The SolutionID with which the nested stack is associated',
            allowedPattern: 'SO([\\d]{4})',
            constraintDescription: 'Please provide the parent stack\'s solutionID. The pattern must match the following regular experssion SO([\d]{4})'
        });

        /**
         * The name of the solution with which the the nested stack is associated with
         */
        const solutionName = new CfnParameter(this, 'SolutionName', {
            type: 'String',
            description: 'The name of the solution for this stack',
            allowedPattern: '[\\w\\W]{1,200}',
            constraintDescription: 'Please provide the name of the solution to which the solution belongs to. The max length of the name is 200 characters'
        });

        /**
         * The name of the parent stack to use the AWS stack name for QuickSight resource creation
         */
        const parentStackName = new CfnParameter(this, 'ParentStackName', {
            type: 'String',
            description: 'The AWS Stack nane of the parent stack',
            allowedPattern: '[\\w\\W]{1,128}',
            constraintDescription: 'Please provide the AWS stack name with which the nested stack is associated with'
        });

        const customResourceRole = new ExecutionRole(this, 'CustomResourceRole');

        this._quickSight = new QuickSight(this, 'Quicksight', {
            name: `${solutionID.valueAsString}-${solutionName.valueAsString}`,
            resource: QuickSightSetup.ALL,
            sourceTemplateArn: quickSightSourceTemplateArn.valueAsString,
            principalArn: quicksightPrincipalArn.valueAsString,
            workgroupName: 'primary',
            logLevel: 'INFO',
            role: customResourceRole.Role,
            parentStackName: parentStackName.valueAsString
        });
    }

    public get analysisURLOutput(): string {
        return this._quickSight.analysisURL;
    }

    public get dashboardURLOutput(): string {
        return this._quickSight.dashboardURL;
    }
}
