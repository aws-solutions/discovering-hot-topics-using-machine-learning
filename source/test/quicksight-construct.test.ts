/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { Template } from 'aws-cdk-lib/assertions';
import { Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { Aws, Stack } from 'aws-cdk-lib';
import { QuickSight, QuickSightSetup } from '../lib/quicksight-custom-resources/quicksight-construct';

test('QS custom resource creation', () => {
    const stack = new Stack();
    new QuickSight(stack, 'testQSConstruct', {
        name: 'solution-name',
        resource: QuickSightSetup.ALL,
        sourceTemplateArn: 'arn:some-parition:quicksight:some-region:fakeaccountid:template/solution_solution-name-v_1_0',
        principalArn: 'arn:some-partition:quicksight:some-region:fakeaccountid:user/namespace/some-role/some-user',
        logLevel: 'INFO',
        workgroupName: 'testGroup',
        role: new Role(stack, 'testRole', {
            assumedBy: new ServicePrincipal('lambda.amazonaws.com')
        }),
        parentStackName: Aws.STACK_NAME
    });
    });