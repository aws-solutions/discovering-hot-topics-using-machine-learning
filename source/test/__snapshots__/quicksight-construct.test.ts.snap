// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QS custom resource creation 1`] = `
{
  "Parameters": {
    "AssetParametersb1ca903e2560d41c92d1b702f7e2f764d7a277170d88d5e14b141a38f4ff3797ArtifactHash29B97AC0": {
      "Description": "Artifact hash for asset "b1ca903e2560d41c92d1b702f7e2f764d7a277170d88d5e14b141a38f4ff3797"",
      "Type": "String",
    },
    "AssetParametersb1ca903e2560d41c92d1b702f7e2f764d7a277170d88d5e14b141a38f4ff3797S3Bucket370C7A2F": {
      "Description": "S3 bucket for asset "b1ca903e2560d41c92d1b702f7e2f764d7a277170d88d5e14b141a38f4ff3797"",
      "Type": "String",
    },
    "AssetParametersb1ca903e2560d41c92d1b702f7e2f764d7a277170d88d5e14b141a38f4ff3797S3VersionKey6859C342": {
      "Description": "S3 key for asset version "b1ca903e2560d41c92d1b702f7e2f764d7a277170d88d5e14b141a38f4ff3797"",
      "Type": "String",
    },
  },
  "Resources": {
    "testQSConstructCustomResource61E8811D": {
      "DependsOn": [
        "testQSConstructQSCustomResourcePolicy4A5596ED",
        "testRole836465CB",
      ],
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W89",
              "reason": "This is not a rule for the general case, just for specific use cases/industries",
            },
            {
              "id": "W92",
              "reason": "Impossible for us to define the correct concurrency for clients",
            },
          ],
        },
      },
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "AssetParametersb1ca903e2560d41c92d1b702f7e2f764d7a277170d88d5e14b141a38f4ff3797S3Bucket370C7A2F",
          },
          "S3Key": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    0,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParametersb1ca903e2560d41c92d1b702f7e2f764d7a277170d88d5e14b141a38f4ff3797S3VersionKey6859C342",
                        },
                      ],
                    },
                  ],
                },
                {
                  "Fn::Select": [
                    1,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParametersb1ca903e2560d41c92d1b702f7e2f764d7a277170d88d5e14b141a38f4ff3797S3VersionKey6859C342",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "lambda_function.handler",
        "Role": {
          "Fn::GetAtt": [
            "testRole836465CB",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "testQSConstructQSCustomResourcePolicy4A5596ED": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "quicksight:CreateAnalysis",
                "quicksight:DeleteAnalysis",
                "quicksight:CreateDataSet",
                "quicksight:DeleteDataSet",
                "quicksight:CreateDataSource",
                "quicksight:DeleteDataSource",
                "quicksight:Describe*",
                "quicksight:Get*",
                "quicksight:List*",
                "quicksight:PassDataSet",
                "quicksight:PassDataSource",
                "quicksight:RestoreAnalysis",
                "quicksight:SearchAnalyses",
                "quicksight:CreateDashboard",
                "quicksight:DeleteDashboard",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":quicksight:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":*/*",
                  ],
                ],
              },
            },
            {
              "Action": "quicksight:DescribeTemplate",
              "Effect": "Allow",
              "Resource": "arn:some-parition:quicksight:some-region:fakeaccountid:template/solution_solution-name-v_1_0",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testQSConstructQSCustomResourcePolicy4A5596ED",
        "Roles": [
          {
            "Ref": "testRole836465CB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testQSConstructQuickSightResourcesA23CC3B0": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "testQSConstructQSCustomResourcePolicy4A5596ED",
      ],
      "Properties": {
        "ApplicationName": "solution-name",
        "LogLevel": "INFO",
        "QuickSightPrincipalArn": "arn:some-partition:quicksight:some-region:fakeaccountid:user/namespace/some-role/some-user",
        "QuickSightSourceTemplateArn": "arn:some-parition:quicksight:some-region:fakeaccountid:template/solution_solution-name-v_1_0",
        "Resource": "all",
        "ServiceToken": {
          "Fn::GetAtt": [
            "testQSConstructCustomResource61E8811D",
            "Arn",
          ],
        },
        "StackName": {
          "Ref": "AWS::StackName",
        },
        "WorkGroupName": "testGroup",
      },
      "Type": "Custom::QuickSightResources",
      "UpdateReplacePolicy": "Delete",
    },
    "testRole836465CB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
